### PR TITLE
[MAINTENANCE] Utilize Rule objects in Profiler construction in DataAssistant

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Use `Rule` objects returned by `DataAssistant.rules()` interface in `Profiler` construction (instead of passing their configurations as overrides to `run()` method) -- this adds flexibility in configuring `DataAssistant` internal `Profiler`.
* Clean up unused imports and outdated `run()` call arguments for `DataAssistant` base class.
* Ensuring that existing tests pass.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-464/GREAT-598/GREAT-822/GREAT-853
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py [hunk 1]
```diff
@@ -150,38 +152,38 @@ def _build_profiler(self) -> None:
...
-                        domain_builder=domain_builder,
-                        parameter_builders=parameter_builders,
-                        expectation_configuration_builders=expectation_configuration_builders,
-                    )
-                )
+            self._add_rules_to_profiler(rules=rules)
+
+        common_rule_names: Set[str] = {
+            rule.name for rule in (self.profiler.rules or {})
+        } & {rule.name for rule in (self.rules or {})}
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: This is super elegant but I'm always going to advocate for simpler code!

I'd ask that we move this logic to a private helper function and actually go through the manual steps of creating two sets, naming them appropriately, and doing the & all in separate lines.

We can even encapsulate the exception checking in that helper (i.e. `_check_for_unique_rules`). When it doubt, KISS 😄 especially as we grow the team and work towards cleaning up the codebase.

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py [hunk 2]
```diff
@@ -150,38 +152,38 @@ def _build_profiler(self) -> None:
...
+        if common_rule_names:
+            raise ge_exceptions.ProfilerConfigurationError(
+                message=f"""Rule names in {self.__class__.__name__} must be unique; duplicate(s) found \
+({common_rule_names}).
+"""
+            )
 
+        self._add_rules_to_profiler(rules=self.rules)
+
+        variables.update(self.variables or {})
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: Same note here. Could we just do a check to see if `self.variables` is present and only update if so? Very minor change but I think it improves readability.

I feel less strongly about this than the prior comment so I'll leave it up to you.

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py [hunk 3]
```diff
@@ -353,6 +355,47 @@ def batch_id_to_batch_identifier_display_name_map(
...
+        """
+        This private utility method adds supplied "Rule" objects to underlying "BaseRuleBasedProfiler" object.
+
+        Args:
+            rules: List of "Rule" objects to be added to given "BaseRuleBasedProfiler" object
+        """
+        rule: Rule
+        domain_builder: DomainBuilder
+        rule_parameter_builders: List[ParameterBuilder]
+        metric_parameter_builders: List[ParameterBuilder]
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: This should be `Optional` no? The `get` call on L378 higlights the possibility of a null value.

File: great_expectations/rule_based_profiler/data_assistant/data_assistant_runner.py [hunk 4]
```diff
@@ -35,21 +34,24 @@ def __init__(
 
     def run(
         self,
-        batch_request: Union[BatchRequestBase, dict],
-        expectation_suite: Optional[ExpectationSuite] = None,
-        expectation_suite_name: Optional[str] = None,
+        variables: Optional[Dict[str, Any]] = None,
+        rules: Optional[Dict[str, Dict[str, Any]]] = None,
+        batch_request: Optional[Union[BatchRequestBase, dict]] = None,
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant_runner.py:39]: Are these not pretty big changes to our API? Are we confident this is the direction to take?

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: @cdkini Good suggestion -- fixing now (in next commit).  Thanks!

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: @cdkini Fixing (in next commit).  Thanks!

File: great_expectations/rule_based_profiler/data_assistant/data_assistant_runner.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant_runner.py:39]: @cdkini No, this is a fix for essentially a typo left behind.  Happy to discuss.

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: @cdkini Made "Optional" (in next commit).  Thanks.

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py [hunk 5]
```diff
@@ -353,6 +349,70 @@ def batch_id_to_batch_identifier_display_name_map(
...
+
+        custom_rule_names: Set[str] = {rule.name for rule in custom_rules}
+
+        common_rule_names: Set[str] = profiler_rule_names & custom_rule_names
+        if common_rule_names:
+            raise ge_exceptions.ProfilerConfigurationError(
+                message=f"""Rule names in {self.__class__.__name__} must be unique; duplicate(s) found \
+({common_rule_names}).
+"""
+            )
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: Thank you 🙇🏽 

Please annotate the return as `-> None`

**Review by cdkini — APPROVED** [review]

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: @cdkini Done (and added a docstring) -- thanks.

**Pull request merged**